src/eduid_webapp/eidas/acs_actions.py

```diff
@@ -188,3 +189,56 @@ def nin_verify_action(session_info, user):
     new_query_string = urlencode({'msg': 'eidas.nin_verify_success'})
     url = urlunsplit((scheme, netloc, path, new_query_string, fragment))
     return redirect(url)
+
+
+@acs_action('mfa-authentication-action')
+@require_user
+def mfa_authentication_action(session_info, user):
+    relay_state = request.form.get('RelayState')
+    current_app.logger.debug('RelayState: {}'.format(relay_state))
+    redirect_url = session.pop(relay_state, None)
+    if not redirect_url:
+        # With no redirect url just redirect the user to dashboard for a new try to log in
+        scheme, netloc, path, query_string, fragment = urlsplit(current_app.config['DASHBOARD_URL'])
+        new_query_string = urlencode({'msg': ':ERROR:eidas.no_redirect_url'})
+        url = urlunsplit((scheme, netloc, path, new_query_string, fragment))
+        return redirect(url)
+
+    # We get the mfa authentication views "next" argument as base64 to avoid our request sanitation
+    # to replace all & to &amp;
+    redirect_url = base64.b64decode(redirect_url).decode('utf-8')
+    # TODO: Rename verify_relay_state to verify_redirect_url
+    redirect_url = verify_relay_state(redirect_url)
+    scheme, netloc, path, query_string, fragment = urlsplit(redirect_url)
+
+    if not is_required_loa(session_info, 'loa3'):
+        new_query_string = urlencode({'msg': ':ERROR:eidas.authn_context_mismatch'})
+        url = urlunsplit((scheme, netloc, path, new_query_string, fragment))
+        return redirect(url)
+
+    if not is_valid_reauthn(session_info):
+        new_query_string = urlencode({'msg': ':ERROR:eidas.reauthn_expired'})
+        url = urlunsplit((scheme, netloc, path, new_query_string, fragment))
+        return redirect(url)
+
+    # Check that a verified NIN is equal to the asserted attribute personalIdentityNumber
+    asserted_nin = get_saml_attribute(session_info, 'personalIdentityNumber')[0]
+    user_nin = user.nins.verified.find(asserted_nin)
+    if not user_nin:
+        current_app.logger.error('Asserted NIN not matching user verified nins')
+        current_app.logger.debug('Asserted NIN: {}'.format(asserted_nin))
+        new_query_string = urlencode({'msg': ':ERROR:eidas.nin_not_matching'})
+        url = urlunsplit((scheme, netloc, path, new_query_string, fragment))
+        return redirect(url)
+
+    # TODO: Create dataclass for session namespace
+    session['mfa_authentication_success'] = True
```

Comment:
maybe it will be easier to evolve into namespaced data later if these four values were in a dict under session['mfa_authentication']?